module_3/gifts/service/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java

```diff
@@ -0,0 +1,338 @@
+		this.orderRepository = orderRepository;
+		this.userRepository = userRepository;
+		this.certificateRepository = certificateRepository;
+		this.orderConverter = orderConverter;
+		this.orderDataConverter = orderDataConverter;
+		this.orderValidation = orderValidation;
+	}
+
+	/**
+	 * Reads order with passed id.
+	 * 
+	 * @param orderId id of the order to be read
+	 * @return order with passed id
+	 * @throws ValidationException if passed order id is invalid
+	 * @throws NotFoundException   if the order with passed id does not exist
+	 */
+	@Override
+	public OrderDto readById(long orderId) {
+		if (!ValidationUtil.isPositive(orderId)) {
+			throw new ValidationException(EntityConstant.ID + ValidationUtil.ERROR_RESOURCE_DELIMITER + orderId,
+					ErrorCode.INVALID_ORDER_ID);
+		}
+
+		Optional<OrderModel> orderModel = orderRepository.findById(orderId);
+
+		if (orderModel.isEmpty()) {
+			throw new NotFoundException(EntityConstant.ID + ValidationUtil.ERROR_RESOURCE_DELIMITER + orderId,
+					ErrorCode.NO_ORDER_FOUND);
+		}
+
+		OrderDto orderDto = orderConverter.convertToDto(orderModel.get());
+
+		return orderDto;
+	}
+
+	/**
+	 * Reads all orders for the specified user according to the passed parameters.
+	 * 
+	 * @param userId id of the user whose orders should be read
+	 * @param params the parameters which define the choice of orders and their
+	 *               ordering
+	 * 
+	 * @return orders for specified user which meet the passed parameters
+	 * @throws NotFoundException   if user with passed id does not exist
+	 * @throws ValidationException if passed user id or read parameters are invalid
+	 */
+	@Override
+	public List<OrderDto> readAllByUserId(long userId, MultiValueMap<String, String> params) {
+		MultiValueMap<String, String> paramsInLowerCase = ValidationUtil.mapToLowerCase(params);
+		if (!ValidationUtil.isPositive(userId)) {
+			throw new ValidationException(EntityConstant.ID + ValidationUtil.ERROR_RESOURCE_DELIMITER + userId,
+					ErrorCode.INVALID_USER_ID);
+		}
+		if (!userRepository.userExistsById(userId)) {
+			throw new NotFoundException(EntityConstant.ID + ValidationUtil.ERROR_RESOURCE_DELIMITER + userId,
+					ErrorCode.NO_USER_FOUND);
+		}
+		Map<ErrorCode, String> errors = orderValidation.validateReadParams(paramsInLowerCase);
+
+		if (!errors.isEmpty()) {
+			throw new ValidationException(errors, ErrorCode.INVALID_ORDER_REQUEST_PARAMS);
+		}
+
+		int offset = ServiceConstant.OFFSET;
+		int limit = ServiceConstant.LIMIT;
+
+		if (params.containsKey(EntityConstant.OFFSET)) {
+			offset = Integer.parseInt(params.get(EntityConstant.OFFSET).get(0));
+		}
+
+		if (params.containsKey(EntityConstant.LIMIT)) {
+			limit = Integer.parseInt(params.get(EntityConstant.LIMIT).get(0));
+		}
+
+		List<OrderModel> orderModels = orderRepository.readAllByUserId(userId, offset, limit);
+		List<OrderDto> orderDtos = new ArrayList<>(orderModels.size());
+		for (OrderModel orderModel : orderModels) {
```

Comment:
as I see you can simply all the similar foreach with stream api in the whole project